File: fastapi_ec2/entity/message.py
import datetime
from enum import Enum
from typing import TYPE_CHECKING, List, Optional

from pgvector.sqlalchemy import Vector
from sqlalchemy import (
    ARRAY,
    BigInteger,
    Boolean,
    CheckConstraint,
    DateTime,
    Double,
    ForeignKey,
    SmallInteger,
    String,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base
from .group import MemberGroup
from .member import Member

if TYPE_CHECKING:
    from .message_box import MessageBox

IMAGE_VECTOR_SIZE = 12288


class VarcharLimit(Enum):
    TITLE: int = 15
    CONTENT: int = 100
    HINT: int = 20


class ReceiverTypes(Enum):
    INDIVIDUAL: int = 0
    GROUP_UNCONSTRUCTED: int = 1
    GROUP_CONSTRUCTED: int = 2
    PUBLIC: int = 3


class Message(Base):
    __tablename__ = "message"
    __table_args__ = (
        CheckConstraint(
            f"receiver_type IN ({ReceiverTypes.INDIVIDUAL.value}, {ReceiverTypes.GROUP_UNCONSTRUCTED.value}, {ReceiverTypes.PUBLIC.value}, {ReceiverTypes.GROUP_CONSTRUCTED.value})",
            name="receiver_type_check",
        ),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    sender_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    receiver_type: Mapped[int] = mapped_column(
        SmallInteger, nullable=False, server_default=f"{ReceiverTypes.INDIVIDUAL.value}"
    )
    receiver: Mapped[Optional[list[int]]] = mapped_column(
        ARRAY(BigInteger), nullable=True
    )
    hint_image_first: Mapped[str] = mapped_column(String(255), nullable=True)
    hint_image_second: Mapped[str] = mapped_column(String(255), nullable=True)
    dot_hint_image: Mapped[str] = mapped_column(String(255), nullable=True)
    title: Mapped[str] = mapped_column(String(VarcharLimit.TITLE.value), nullable=False)
    content: Mapped[Optional[str]] = mapped_column(
        String(VarcharLimit.CONTENT.value), nullable=True
    )
    hint: Mapped[Optional[str]] = mapped_column(
        String(VarcharLimit.HINT.value), nullable=True
    )
    lat: Mapped[float] = mapped_column(Double, nullable=True)
    lng: Mapped[float] = mapped_column(Double, nullable=True)
    coordinate: Mapped[list[float]] = mapped_column(Vector(3), nullable=True)
    is_treasure: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default="false"
    )
    is_found: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default="false"
    )
    created_at: Mapped[datetime.datetime] = mapped_column(DateTime, nullable=False)
    image: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    vector: Mapped[list[float]] = mapped_column(
        Vector(IMAGE_VECTOR_SIZE), nullable=True
    )
    group_id: Mapped[Optional[int]] = mapped_column(
        BigInteger, ForeignKey("member_group.id", ondelete="SET NULL"), nullable=True
    )
    member: Mapped[Optional["Member"]] = relationship(
        "Member",
        primaryjoin="foreign(Message.sender_id) == Member.id",
        viewonly=True,
    )
    group: Mapped[Optional["MemberGroup"]] = relationship(
        "MemberGroup",
        viewonly=True,
    )
    message_boxes: Mapped[Optional[List["MessageBox"]]] = relationship(
        "MessageBox", viewonly=True, uselist=True
    )

    def __repr__(self) -> str:
        return f"Message(id={self.id}, hint_image={self.hint_image_first}, wrote_by={self.sender_id})"

File: star/src/main/java/com/ssafy/star/message/entity/Message.java
package com.ssafy.star.message.entity;

import com.ssafy.star.member.entity.Member;
import com.ssafy.star.member.entity.MemberGroup;
import io.hypersistence.utils.hibernate.type.array.ListArrayType;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Type;

import java.time.LocalDateTime;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "message")
public class Message {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "sender_id", nullable = false)
    private Member sender;

    @Column(name="receiver_type", nullable = false)
    private short receiverType = 0;

    @Column(name = "hint_image_first")
    private String hintImageFirst;
    @Column(name = "hint_image_second")
    private String hintImageSecond;
    @Column(name = "dot_hint_image")
    private String dotHintImage;

    @Column(nullable = false)
    private String title;

    @Column
    private String content;

    @Column
    private String hint;

    @Column
    private Float lat;
    @Column
    private Float lng;
//    @Column
//    @JdbcTypeCode(SqlTypes.VECTOR)
//    @Array(length = 3)
//    private Float[] coordinate;

    @Column(name = "is_treasure", nullable = false)
    private boolean isTreasure = false;

    @Column(name = "is_found", nullable = false)
    private boolean isFound = false;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column
    private String image;

//    @Column
//    @JdbcTypeCode(SqlTypes.VECTOR)
//    @Array(length = 12288)
//    private Float[] vector;

    @ManyToOne
    @JoinColumn(name = "group_id", foreignKey = @ForeignKey(name = "fk_message_group"))
    private MemberGroup group;

    @Type(ListArrayType.class)
    @Column(
            name = "receiver",
            columnDefinition = "bigint[]"
    )
    private List<Long> receiver;

}
